### PR TITLE
initial push for hierarcy based permissions

### PR DESCRIPTION
**Pull request by natlibfi-jonollil**

### CODE REVIEW

File: src/interfaces/utils.js [hunk 1]
```diff
@@ -25,14 +25,33 @@
...
+};
 
-export function hasPermission(profile, user) {
-	const permitted = profile.auth.groups.some(profileGroup => {
-		return user.groups.some(userGroup => userGroup === profileGroup || userGroup === 'admin');
-	});
-	return permitted;
-}
+export function hasPermission(type, command, profile, bgroups = []) {
+	const permintted = profile.groups.some(
```
Comment (ghost) [inline src/interfaces/utils.js]: Fix typo: `permintted` -> `permitted`

File: src/interfaces/utils.js [hunk 2]
```diff
@@ -25,14 +25,33 @@
...
+	}
+};
 
-export function hasPermission(profile, user) {
-	const permitted = profile.auth.groups.some(profileGroup => {
-		return user.groups.some(userGroup => userGroup === profileGroup || userGroup === 'admin');
-	});
-	return permitted;
-}
+export function hasPermission(type, command, profile, bgroups = []) {
```
Comment (ghost) [inline src/interfaces/utils.js]: Change parameter names to be more clear:

- *profile* -> **userGroups**
- *bgroups* -> **permittedGroups**

Also, it's unecessary to pass the whole profile object here. Passing `profile.auth.groups` is sufficient.

File: src/interfaces/utils.js [hunk 3]
```diff
@@ -25,14 +25,33 @@
...
+		group => permissions[type][command].includes(group) ||
+		permissions[type][command].includes('all')
+	);
+	if (profile.groups.includes('system') || bgroups === []) {
+		return permintted;
+	}
 
-export function hasAdminPermission(user) {
-	return hasPermission({auth: {groups: ['admin']}}, user);
+	return profile.groups.some(group => bgroups.includes(group)) && permintted;
```
Comment (ghost) [inline src/interfaces/utils.js]: Optimization: Turn this the other way around nt reduce unnecessary array iteration if the boolean doesn't match:
`return permitted  && profile.groups.some(group => bgroups.includes(group))`

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Review by ghost — CHANGES_REQUESTED** [review]
Doesn't seem to check permission to use the `abort` operation of updateBlob?

**Comment by natlibfi-jonollil** [comment]
> Doesn't seem to check permission to use the `abort` operation of updateBlob?

It needed `UPDATE` tier permissions to `ABORT`blob process. Fixed, updated tests and tested.

File: src/interfaces/utils.js [hunk 4]
```diff
@@ -45,13 +45,24 @@ const permissions = {
 };
 
 export function hasPermission(type, command, userGroups, permittedGroups = []) {
```
Comment (ghost) [inline src/interfaces/utils.js:47]: Make this more easier to read:
```js
const commandPermissions = permissions[type][command];

if (userGroups.includes('system')) {
  return true;

if (hasGroup(permittedGroups) === false) {
  return false;
}

if (commandPermissions.includes('all') || hasGroup(commandPermissions)) {
  return true;
}

function hasGroup(permitted) {
  return userGroups.some(g => permitted.includes(g));
}
```

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by natlibfi-jonollil** [comment]
https://github.com/NatLibFi/melinda-record-import-api/pull/14 merge first if possible?

**Review by ghost — APPROVED** [review]

**Pull request merged**